File: golang/data_structures/matrices/spiral_matrix_3/spiral_matrix_3.go
package main

import (
	"fmt"
)

func spiralOrder(rows int, cols int, rStart int, cStart int) [][]int {
	directions := [][]int{
		{0, 1},  // right
		{1, 0},  // down
		{0, -1}, // left
		{-1, 0}, // up
	}
	currentDirection := 0
	steps := 1
	r, c := rStart, cStart
	result := [][]int{}
	visited := make(map[string]bool)

	for len(result) < rows * cols {
		for i := 0; i < steps; i++ {
			if r >= 0 && r < rows && c >= 0 && c < cols && !visited[fmt.Sprintf("%d,%d", r, c)] {
				result = append(result, []int{r, c})
				visited[fmt.Sprintf("%d,%d", r, c)] = true
			}
			r += directions[currentDirection][0]
			c += directions[currentDirection][1]
		}

		currentDirection = (currentDirection + 1) % 4

		if currentDirection == 0 || currentDirection == 2 {
			steps++
		}
	}

	return result
}

func main() {
	fmt.Println(spiralOrder(5, 6, 1, 4))
}

File: typescript/data_structures/matrices/spiral_matrix_3.ts
function spiralMatrixIII(rows: number, cols: number, rStart: number, cStart: number): number[][] {
    const result: number[][] = []
    const visited = new Set<string>()
    const directions = [
        [0, 1],  // right
        [1, 0],  // down
        [0, -1], // left
        [-1, 0]  // up
    ]
    let currentDirection = 0
    let steps = 1
    let r = rStart
    let c = cStart

    while (result.length < rows * cols) {
        for (let i = 0; i < steps; i++) {
            if (r >= 0 && r < rows && c >= 0 && c < cols && !visited.has(`${r},${c}`)) {
                result.push([r, c])
                visited.add(`${r},${c}`)
            }
            r += directions[currentDirection][0]
            c += directions[currentDirection][1]
        }

        currentDirection = (currentDirection + 1) % 4

        if (currentDirection === 0 || currentDirection === 2) {
            steps++
        }
    }

    return result
}

console.log(spiralMatrixIII(5, 6, 1, 4))
